File: Java/dependency-injection/bastard-injection/src/main/java/com/github/grzesiek_galezowski/BastardInjection/Core/AcmeProcessingWorkflow.java
package com.github.grzesiek_galezowski.BastardInjection.Core;

import com.github.grzesiek_galezowski.BastardInjection.Interfaces.AcmeMessage;
import com.github.grzesiek_galezowski.BastardInjection.Outbound.IOutbound;
import com.github.grzesiek_galezowski.BastardInjection.Services.ActiveDirectoryBasedAuthorization;
import com.github.grzesiek_galezowski.BastardInjection.Services.IAuthorization;
import com.github.grzesiek_galezowski.BastardInjection.Services.IRepository;
import com.github.grzesiek_galezowski.BastardInjection.Services.MsSqlBasedRepository;

class AcmeProcessingWorkflow implements IProcessingWorkflow {
  private final IRepository _repository;
  private final IAuthorization _authorizationRules;
  private IOutbound _outbound;

  public AcmeProcessingWorkflow() {
    this(new ActiveDirectoryBasedAuthorization(), new MsSqlBasedRepository());
  }

  //for tests
  public AcmeProcessingWorkflow(
      IAuthorization authorization,
      IRepository repository) {
    _authorizationRules = authorization;
    _repository = repository;
  }

  public void setOutbound(IOutbound outbound) {
    _outbound = outbound;
  }

  public void applyTo(AcmeMessage message) {
    message.authorizeUsing(_authorizationRules);
    _repository.save(message);
    _outbound.send(message);
  }
}

File: CSharp/DependencyInjection/BastardInjection/_4_BastardInjection/Core/AcmeProcessingWorkflow.cs
using BastardInjection._4_BastardInjection.Interfaces;
using BastardInjection._4_BastardInjection.Outbound;
using BastardInjection._4_BastardInjection.Services;

namespace BastardInjection._4_BastardInjection.Core
{
  internal interface IProcessingWorkflow
  {
    void SetOutbound(IOutbound outbound);
    void ApplyTo(AcmeMessage message);
  }

  class AcmeProcessingWorkflow : IProcessingWorkflow
  {
    private readonly IRepository _repository;
    private readonly IAuthorization _authorizationRules;
    private IOutbound _outbound;

    public AcmeProcessingWorkflow() : this(new ActiveDirectoryBasedAuthorization(), new MsSqlBasedRepository())
    {
    }

    //for tests
    private AcmeProcessingWorkflow(IAuthorization authorization, IRepository repository)
    {
      _authorizationRules = authorization;
      _repository = repository;
    }

    public void SetOutbound(IOutbound outbound)
    {
      _outbound = outbound;
    }

    public void ApplyTo(AcmeMessage message)
    {
      message.AuthorizeUsing(_authorizationRules);
      _repository.Save(message);
      _outbound.Send(message);
    }
  }
}